artifacts/reader.py

```diff
@@ -167,59 +170,65 @@ def ReadArtifactDefinition(self, artifact_definition):
+  def _ReadLabels(self, artifact_definition_values, artifact_definition):
     """Reads the optional artifact definition labels.
 
     Args:
-      artifact_definition: the artifact definition as a dict.
-      artifact_object: the artifact definition object (instance of
+      artifact_definition_values: dictionary containing the artifact definition
+                                  values.
+      artifact_definition: the artifact definition (instance of
                            ArtifactDefinition).
+
     Raises:
       FormatError: if there are undefined labels.
     """
-    name = artifact_definition.get(u'name', None)
-    labels = artifact_definition.get(u'labels', [])
-    undefined_labels = [item for item in labels if item not in self.labels]
+    name = artifact_definition_values.get(u'name', None)
+    labels = artifact_definition_values.get(u'labels', [])
 
+    undefined_labels = set(labels).difference(self.labels)
     if undefined_labels:
       raise errors.FormatError(
-          u'Artifact definition: {0} label(s): {1} not defined.'.format(
-              name, ', '.join(undefined_labels)))
+          u'Artifact definition: {0} found undefined labels: {1}.'.format(
+              name, u', '.join(undefined_labels)))
 
-    artifact_object.labels = artifact_definition.get(u'labels', [])
+    artifact_definition.labels = labels
 
-  def _ReadSupportedOS(self, definition, definition_object, name):
+  def _ReadSupportedOS(self, definition_values, definition_object, name):
     """Reads the optional artifact or source type supported OS.
+
     Args:
-      definition: the artifact or source definition as a dict.
+      definition_values: a dictionary containing the artifact defintion or
+                         source type values.
       definition_object: the definition object (instance of ArtifactDefinition
-                        or SourceType).
+                         or SourceType).
       name: string containing the name of the artifact definition.
+
     Raises:
       FormatError: if there are undefined supported operating systems.
     """
-    supported_os = definition.get(u'supported_os', [])
-
+    supported_os = definition_values.get(u'supported_os', [])
     if not isinstance(supported_os, list):
-      raise errors.FormatError(
-          u'supported_os must be a list of strings, got: {0}'.format(
-              supported_os))
-
-    undefined_supported_os = [
-        item for item in supported_os if item not in self.supported_os
-    ]
+      raise errors.FormatError(u'Invalid supported_os type: {0}'.format(type(
+          supported_os)))
 
+    undefined_supported_os = set(supported_os).difference(self.supported_os)
     if undefined_supported_os:
-      error_string = (
-          u'Artifact definition: {0} supported operating system: {1} '
-          u'not defined.').format(name, u', '.join(undefined_supported_os))
+      if isinstance(definition_object, artifact.ArtifactDefinition):
```

Comment:
This seems overly complicated. The reality is you already have the name of the artifact and the string of the undefined OS, which is plenty to track it down quickly. It doesn't really matter what level it is at inside the artifact.